### ISSUE TITLE
[Feature Request] Make `str()` and `repr()` work with all structs, types, traits and classes

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [modular/mojo#1436]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

I would like that I can call `str()` and `repr()` unconditionally on all mojo objects, being structs, traits, types, classes, etc, even the ones not defined by me or those which don't explicitely define a `__str__` or `__repr__`.

### What is your motivation for this change?

Currently, we need to use traits to make sure that we can call `str()` or `repr()` on an struct. While this is useful, Python has opted to have all objects have a default `str()` and `repr()` representation. Why is it useful in python? Multiple reasons:

1) Users, when they learn the language, use and abuse `print()`, and `print()` in python works with all objects, providing a good UX for beginners, and good debuggability.

2) It encourages user-defined structs to define `__str__` and `__repr__`, because users are certain that the objects they hold can be called with `str()` or `repr()`, making it easy to write. This is useful for collections like list, dict, etc... They don't need to add a condition on the type being `Stringable`. One could say that in Python, eveything is Stringable. The fact that custom structs have custom `__str__` and `__repr__` make them easier to debug. We should encourage users to write those.

3) Error messages. When writing error messages in functions, if a user wants to provide information about the object, the user firsts needs to make sure that the object is Stringable. In Python, such a check is not needed, making writing error messages painless and actually enjoyable. The simplest example would be reimplementing Python's `dict` in Mojo, and providing a `KeyError`. A `KeyError` in Python prints the key representation, in mojo we would need to make sure the dictionary keys are `Stringable` first, which may not be the case, thus making the error logic more complicated than in Python.

**Comment by rarebreed** [comment]
In rust, this is possible, as long as you `#[derive(Debug)]` your struct or enum, and you need a special marker in the `println!` to use `{:?}` (or `{:#?}` for pretty printing).

Since traits are in their infancy, mojo may eventually add something like rust's `derive` or haskell's `deriving` feature to auto-implement a trait for a struct (and whatever they will have for a Sum type)

**Comment by rarebreed** [comment]
I forgot to mention, there is a penalty for this...speeding up compile time for the derives.  Which also makes me wonder how mojo will handle traits and generics.  Will mojo do monomorphization like rust does?  Will it use a vtable or dynamic look up to handle polymorphic access with traits? (ie, in rust, there's a difference between `dyn Traits` which use dynamic look up, and regular Traits which get compiled down to type-specific implementations).

**Comment by lattner** [comment]
Yes, I'd like to see this too.  Right now all trait conformances are explicit-only, and we need to decide if we support implicit conformances - the likely answer is "yes".  If so, we could compose that plus some for of synthesized trait magic to generate these things on demand.

**Comment by laszlokindrat** [comment]
Thank you for opening this ticket! We discussed this with the team, and agreed that eventually we would like something like this in Mojo. There are a few dependencies we need to work through before we can do this properly (e.g. improvements to traits, design of existentials), but we agreed that a feature like this would be useful.

**Comment by gryznar** [comment]
@laszlokindrat Implicit conformance is really cool, but I would suggest to restrict it only to certain traits to reduce verbosity. I can easily imagine situation in which accidential convergence in signatures would lead to strange behavior

**Comment by laszlokindrat** [comment]
>We discussed this with the team, and agreed that eventually we would like something like this in Mojo.

Sorry it wasn't clear: I meant to say this about the idea in the OP, i.e. making (almost) everything Stringable. But either way, I agree that implicit conformance is one component we need to think about before we can go there.

**Comment by conorbronsdon** [comment]
[Writable](https://docs.modular.com/mojo/std/format/Writable/) now provides unified string formatting for all types, replacing both Stringable and Representable ([both deprecated](https://docs.modular.com/mojo/changelog/#v0262-2026-03-19)). The key piece is the reflection-based default implementation added in [v26.1](https://docs.modular.com/mojo/changelog/#v0261-2026-01-29), so types get a reasonable string representation automatically without the author having to write one ([source](https://github.com/modular/modular/tree/main/mojo/stdlib/std/format)). Thanks for pushing for this!
